File: second-minimum-node-in-a-binary-tree.cpp
//Runtime: 3 ms
class Solution {
public:

    void inorder(TreeNode* root, set<int>&nums)
    {
        if(!root)
            return;

        inorder(root->left, nums);
        nums.insert(root->val);
        inorder(root->right, nums);
    }

    int findSecondMinimumValue(TreeNode* root) {
        set<int> nums;
        inorder(root, nums);

        if(nums.size() < 2)
            return -1;

        return *(++nums.begin());
    }
};

File: second-minimum-node-in-a-binary-tree.scala
//Runtime: 264 ms
object Solution {

    def findSecondMinimumValue(root: TreeNode): Int = {
        def inorder(root: TreeNode): Set[Int] = root match{
            case null => Set()
            case r: TreeNode => inorder(r.left) ++ Set(r.value) ++ inorder(r.right)
        }
        inorder(root).toList.sortWith(_<_) match{
            case Nil => -1
            case h :: Nil => -1
            case h :: tail=> tail.head
        }
    }
}
